codeview/src/main/java/com/amrdeveloper/codeview/UndoRedoManager.java

```diff
@@ -0,0 +1,267 @@
+        editText = edittext;
+        mEditHistory = new EditHistory();
+        mChangeListener = new EditTextChangeListener();
+        editText.addTextChangedListener(mChangeListener);
+    }
+
+    public void disconnect() {
+        editText.removeTextChangedListener(mChangeListener);
+    }
+
+    public void setMaxHistorySize(int maxHistorySize) {
+        mEditHistory.setMaxHistorySize(maxHistorySize);
+    }
+
+    public void clearHistory() {
+        mEditHistory.clear();
+    }
+
+    public boolean canUndo() {
+        return (mEditHistory.mmPosition > 0);
+    }
+
+    public void undo() {
+        EditItem edit = mEditHistory.getPrevious();
+        if (edit == null) {
+            return;
+        }
+
+        Editable text = editText.getEditableText();
+        int start = edit.mmStart;
+        int end = start + (edit.mmAfter != null ? edit.mmAfter.length() : 0);
+
+        mIsUndoOrRedo = true;
+        text.replace(start, end, edit.mmBefore);
+        mIsUndoOrRedo = false;
+        for (Object o : text.getSpans(0, text.length(), UnderlineSpan.class)) {
+            text.removeSpan(o);
+        }
+
+        Selection.setSelection(text, edit.mmBefore == null ? start
+                : (start + edit.mmBefore.length()));
+    }
+
+    public boolean canRedo() {
+        return (mEditHistory.mmPosition < mEditHistory.mmHistory.size());
+    }
+
+    public void redo() {
+        EditItem edit = mEditHistory.getNext();
+        if (edit == null) {
+            return;
+        }
+
+        Editable text = editText.getEditableText();
+        int start = edit.mmStart;
+        int end = start + (edit.mmBefore != null ? edit.mmBefore.length() : 0);
+
+        mIsUndoOrRedo = true;
+        text.replace(start, end, edit.mmAfter);
+        mIsUndoOrRedo = false;
+        for (Object o : text.getSpans(0, text.length(), UnderlineSpan.class)) {
+            text.removeSpan(o);
+        }
+
+        Selection.setSelection(text, edit.mmAfter == null ? start
+                : (start + edit.mmAfter.length()));
+    }
+
+    public void storePersistentState(Editor editor, String prefix) {
+        editor.putString(prefix + ".hash",
+                String.valueOf(editText.getText().toString().hashCode()));
+        editor.putInt(prefix + ".maxSize", mEditHistory.mmMaxHistorySize);
+        editor.putInt(prefix + ".position", mEditHistory.mmPosition);
+        editor.putInt(prefix + ".size", mEditHistory.mmHistory.size());
+
+        int i = 0;
+        for (EditItem ei : mEditHistory.mmHistory) {
+            String pre = prefix + "." + i;
+
+            editor.putInt(pre + ".start", ei.mmStart);
+            editor.putString(pre + ".before", ei.mmBefore.toString());
+            editor.putString(pre + ".after", ei.mmAfter.toString());
+
+            i++;
+        }
+    }
+
+    public boolean restorePersistentState(SharedPreferences sp, String prefix)
+            throws IllegalStateException {
```

Comment:
Why we need `throws IllegalStateException`? btw `doRestorePersistentState` didn't throw any exception